### PR TITLE
DEV-2615: Fix paste dropping columns for a ragged clipboard (#7389)

### PR DESCRIPTION
**Pull request by budnix**
### Context

The PR fixes silent data loss when pasting a **ragged** clipboard — one whose first row holds fewer cells than a later row. The surplus columns were dropped with no error and no warning.

Reported in #7389, open since 2020-11-15 and confirmed still reproducible on `develop` @ `e7fa02fdff`.

The payload from the issue:

```
schedule
<TAB>Football<TAB>Score
```

Handsontable parses that to `[["schedule"], ["", "Football", "Score"]]`, then decides how many columns to write from the length of **row 0** alone, and gets `1`. `Football` and `Score` are never written anywhere. Google Sheets pads the short rows and pastes all three columns.

Two independent places made the same mistake:

1. **Plain-text paste** — `copyPaste.ts` read `plainData[0].length` into `populatedColumnsLength`, which bounds both the column loop and its modulo, so any cell past row 0's width was unreachable. `SheetClip.parse` does not pad, so the ragged array arrives intact.
2. **HTML paste** — `parseTable.ts` measured `countCols` from the first `<tr>`, so `htmlToGridSettings` allocated a grid that was already too narrow.

The second defect is the more damaging one and is not mentioned in the issue thread. It discards the cells **before** the `beforePaste` hook runs, so integrators cannot repair it — whereas defect 1 runs after the hook, which is why padding rows in `beforePaste` worked as a workaround for plain text but never for HTML. Real Excel and Google Sheets pastes carry a `text/html` table and Handsontable prefers that flavor, so the path with no workaround was the common one.

Both now take the width from the **widest** row. A short row is padded out and blanks the cells it covers — which is already what happened whenever a row other than the first was the short one, so the rule is now applied uniformly rather than changed.

Two things that make this bug easy to mis-reproduce, noted here for reviewers:

- The **reverse** shape (first row widest) was never broken — short rows read `undefined` and get blanks. A repro built that way shows green.
- Paste **does** grow the grid's column count for a rectangular array (a 3-wide paste into a 2-column grid yields 2x3), so "the grid is too narrow" is not the cause.

### How has this been tested?

Added coverage, and verified each new test fails without the fix and passes with it (the fix was reverted, and the pre-fix bundle swapped in, to confirm).

- **Unit** — 4 cases in `handsontable/src/utils/__tests__/parseTable.unit.ts` covering a narrow first row, a widest row in the middle, a `colspan` in a later row, and a guard that a first-row-widest table is unchanged. The first three fail without the fix.

  ```bash
  npm run test:unit --prefix handsontable -- --testPathPattern=parseTable
  ```

- **E2E (Playwright, real clipboard)** — 3 cases in `tests/e2e/clipboard.spec.ts` for ragged plain text, a ragged HTML table, and the first-row-widest guard. The HTML case deliberately carries a different `text/plain` flavor so the assertions can only pass if the HTML path was the one consumed. Two new `GridPage` helpers write the real clipboard.

  ```bash
  cd tests && npx playwright test e2e/clipboard.spec.ts --project=e2e-main
  ```

Full suites run green before the commit:

- `npm run test:unit --prefix handsontable` — 3736 passed, 312 suites
- `npm run test:e2e --prefix handsontable -- --testPathPattern=copyPaste` — 249 specs, 0 failures
- `npm run test:e2e --prefix handsontable -- --testPathPattern=parseTable` — 11 specs, 0 failures
- `cd tests && npx playwright test --project=e2e-main` — 400 passed, 1 pre-existing skip
- `npm run eslint --prefix handsontable` — 0 errors
- `npm run test:types --prefix handsontable` — passed

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):

1. DEV-2615
2. #7389
3. Possibly related, reported alongside it: #7361

### Affected project(s):

- [x] `handsontable`
- [ ] `@handsontable/angular-wrapper`
- [ ] `@handsontable/react-wrapper`
- [ ] `@handsontable/vue3`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://cla.handsontable.com/sign) — one signature covers both Handsontable and HyperFormula; the `cla/signed` check on this PR confirms it.
- [ ] My change requires a change to the documentation.

ClickUp task: https://app.clickup.com/t/9015210959/DEV-2615

---

> [!NOTE]
> **Medium Risk**
> Changes core paste paths (plain and HTML) and what `beforePaste` receives for ragged input; behavior is intentional and heavily tested but integrators that assumed first-row width or ragged hook payloads may need to adjust.
> 
> **Overview**
> Fixes **silent column loss** when pasting clipboard rows of unequal length (e.g. a narrow first row followed by wider rows, as in #7389). Paste width is now driven by the **widest row**, matching spreadsheet behavior.
> 
> **Plain-text paste** (`copyPaste.ts`): column count and tiling use the max row length instead of the first row. Short rows are padded with **`null`** (not `undefined`, so object data sources stay serializable). **`padRowsToWidest`** runs before **`beforePaste`**, so hook data matches what the grid writes.
> 
> **HTML paste** (`parseTable.ts`): table width comes from **`countTableColumns`** across all rows (colspan/rowspan, last-`td` rules), not the first `<tr>`. Full-width footer colspans no longer inflate width; spans past the last column are **trimmed** when filling the grid.
> 
> Documentation adds a **“Rows of unequal length”** section on the clipboard guide. Coverage includes new unit tests, expanded `parseTable` cases, Playwright ragged paste scenarios, and **`GridPage`** clipboard write helpers.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 11d256e5ec9d2b47ef1cd513ab67f86d9e4762c5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qunabu** [comment]
Task linked: [DEV-2615 Fix paste dropping columns when the clipboard's first row is the narrowest (#7389)](https://app.clickup.com/t/9015210959/DEV-2615)

File: handsontable/src/utils/parseTable.ts [hunk 1]
```diff
@@ -260,6 +260,44 @@ export function replaceTdCellsWithTextContent(html: string): string {
...
+
+  for (let row = 0; row < rows.length; row += 1) {
+    const cells = rows[row].cells;
+    let cols = reserved[row] || 0;
+
+    for (let cell = 0; cell < cells.length; cell += 1) {
+      const { colSpan, rowSpan, nodeName } = cells[cell];
+
+      if (!(hasRowHeaders && cell === 0 && nodeName === 'TH')) {
+        cols += colSpan;
```
Comment (Haxikowy) [inline handsontable/src/utils/parseTable.ts]: `cols += colSpan` lets a row whose width comes only from a spanning cell set the width for the whole table.

```js
htmlToGridSettings('<table><tbody><tr><td>A1</td><td>B1</td><td>C1</td></tr><tr><td colspan="20">footer</td></tr></tbody></table>').data[0].length
// branch: 20   develop: 3
```

`colspan="1000"` (the DOM ceiling) gives a 1000-column grid and a rows x 1000 array. A table with a full-width footer row, normal in email and Word exports, now adds 17 blank columns and blanks 17 cells in every pasted row. Develop pasted that same clipboard cleanly, so this is a new instance of the silent overwrite the PR is fixing.

The ragged-rows argument in the description does not cover it: no row in that table holds 20 discrete cells. A colspan-only row shouldn't widen the count past the widest row of discrete cells. Needs a test for colspan overflow either way.

File: handsontable/src/utils/parseTable.ts [hunk 2]
```diff
@@ -260,6 +260,44 @@ export function replaceTdCellsWithTextContent(html: string): string {
...
+
+        for (let spanned = row + 1; spanned < row + rowSpan; spanned += 1) {
```
Comment (Haxikowy) [inline handsontable/src/utils/parseTable.ts]: Unbounded by the table's own row count, so the loop writes `reserved` entries nothing reads. Measured in Chrome: a 2-row table with 1000 `rowspan="65534"` cells takes 151.57 ms, clamped 0.13 ms, same result. 10000x20 with `rowspan="10000"` on every cell: 4579 ms, clamped 779 ms. Span-free 10000x20 is unchanged (12.66 vs 12.42 ms), so the clamp is free.

```suggestion
        const spannedEnd = Math.min(row + rowSpan, rows.length);

        for (let spanned = row + 1; spanned < spannedEnd; spanned += 1) {
```

File: handsontable/src/utils/parseTable.ts [hunk 3]
```diff
@@ -260,6 +260,44 @@ export function replaceTdCellsWithTextContent(html: string): string {
...
+  let maxCols = 0;
+
+  for (let row = 0; row < rows.length; row += 1) {
+    const cells = rows[row].cells;
+    let cols = reserved[row] || 0;
+
+    for (let cell = 0; cell < cells.length; cell += 1) {
+      const { colSpan, rowSpan, nodeName } = cells[cell];
+
+      if (!(hasRowHeaders && cell === 0 && nodeName === 'TH')) {
```
Comment (Haxikowy) [inline handsontable/src/utils/parseTable.ts]: The exemption skips a `th` only at index 0, but the fill loop sends every non-`TD` cell to `rowHeaders` and gives it no slot. A `th` anywhere else is counted and leaves a phantom trailing column: `<tr><td>a</td><th>grp</th><td>b</td></tr><tr><td>1</td><td>2</td></tr>` parses as `[['a','b',undefined],['1','2',undefined]]`. Develop had the same defect but only for row 0. `nodeName !== 'TD'` would match the fill loop.

File: handsontable/src/utils/parseTable.ts [hunk 4]
```diff
@@ -292,9 +330,7 @@ export function htmlToGridSettings(element: HTMLTableElement | string, rootDocum
   const el: HTMLTableElement = checkElement as HTMLTableElement;
   const generator = tempElem.querySelector('meta[name$="enerator"]') as HTMLMetaElement | null;
   const hasRowHeaders = el.querySelector('tbody th') !== null;
-  const trElement = el.querySelector('tr') as HTMLTableRowElement | null;
-  const countCols = !trElement ? 0 : (Array.from(trElement.cells)
-    .reduce((cols: number, cell: HTMLTableCellElement) => cols + cell.colSpan, 0)) - (hasRowHeaders ? 1 : 0);
+  const countCols = countTableColumns(el, hasRowHeaders);
```
Comment (Haxikowy) [inline handsontable/src/utils/parseTable.ts]: `countCols` now comes from every row while `colHeaders` and `nestedHeaders` still come from `thead`, so the two can disagree:

```js
htmlToGridSettings('<table><thead><tr><th>h1</th><th>h2</th></tr></thead><tbody><tr><td>a</td></tr><tr><td>a</td><td>b</td><td>c</td></tr></tbody></table>')
// branch: data 3 wide, colHeaders ['h1','h2']   develop: data 2 wide, colHeaders ['h1','h2']
```

They matched by construction before. `htmlToGridSettings` is public (`Handsontable.helper`, type-tested in `test/types`), so a scraped ragged table fed to `new Handsontable` now gets auto-named surplus columns. Padding `colHeaders` and each `nestedHeaders` row out to `countCols` once it is known would keep them in step.

File: handsontable/src/plugins/copyPaste/copyPaste.ts [hunk 5]
```diff
@@ -680,7 +683,9 @@ export class CopyPaste extends BasePlugin {
           cellValue = parsedCellValue;
         }
 
-        newRow.push(cellValue);
+        // A row shorter than the widest one has no value here. Write the empty-cell value rather
+        // than `undefined`, which would delete the property outright in an object data source.
+        newRow.push(cellValue === undefined ? null : cellValue);
```
Comment (Haxikowy) [inline handsontable/src/plugins/copyPaste/copyPaste.ts:708]: `null` is the right value here, it matches what `emptySelectedCells` writes. Two things ride along that the description doesn't mention.

The coercion also fires for a short *later* row, the shape the description calls never broken. An object data source that used to lose those keys now keeps them: develop `{"a":"w"}`, branch `{"a":"w","b":null,"c":null}`, and `getDataAtCell` goes from `undefined` to `null`. Anyone testing `value === undefined` to mean "nothing pasted" sees a change.

The coercion also runs after `beforePaste`, so hooks still see the ragged array on the plain-text path and `undefined` holes on the HTML path while the grid gets `null`s. An integrator syncing from `afterPaste` pushes the ragged array and misses the cells the grid just blanked, which is this PR's own bug one layer out. Padding `pastedData` to the widest row before `runHooks('beforePaste')` fixes both flavors and lets this coercion go away.

File: handsontable/src/plugins/copyPaste/copyPaste.ts [hunk 6]
```diff
@@ -612,7 +612,10 @@ export class CopyPaste extends BasePlugin {
 
     const selection = this.hot.getSelectedRangeActive();
     const populatedRowsLength = plainData.length;
-    const populatedColumnsLength = plainData[0].length;
+    // A ragged clipboard (rows of unequal length) must not be narrowed to the first row's width -
+    // cells past that width would never be written. The widest row wins, as in spreadsheet apps.
+    const populatedColumnsLength = plainData
```
Comment (Haxikowy) [inline handsontable/src/plugins/copyPaste/copyPaste.ts:637]: This also changes the repeat-fill period. Pasting `x\ny\tz\tw` into a 6-wide selection: develop filled row 0 with `x,x,x,x,x,x`, the branch gives `x,null,null,x,null,null`. That's the spreadsheet behavior and a real win, but nothing tests paste-into-a-larger-selection with a ragged clipboard, and the description doesn't mention it.

File: .changelogs/7389.json [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{
+  "issuesOrigin": "public",
+  "title": "Fixed pasted columns being dropped when the clipboard's first row held fewer cells than a later row.",
```
Comment (Haxikowy) [inline .changelogs/7389.json:3]: The title names only the dropped columns. Three user-visible changes aren't in it: short rows now blank the cells they cover, padded object fields are written as `null` instead of being deleted, and repeat-fill into a larger selection tiles on the widest row. `.ai/BREAKING-CHANGES.md` counts a behavior change as declarable, so either extend the title or add a second `changed` entry.

File: tests/e2e/clipboard.spec.ts [hunk 8]
```diff
@@ -52,4 +52,56 @@ test.describe('clipboard', () => {
...
+      // The reverse shape was never broken — this guards it against the fix.
+      await grid.selectCell(0, 0);
+      await grid.writeClipboardText('x\ty\tz\nw');
+      await page.keyboard.press('ControlOrMeta+v');
+
+      await grid.expectCell(0, 0, 'x');
+      await grid.expectCell(0, 1, 'y');
+      await grid.expectCell(0, 2, 'z');
+      await grid.expectCell(1, 0, 'w');
+      await grid.expectCell(1, 1, '');
```
Comment (Haxikowy) [inline tests/e2e/clipboard.spec.ts]: `expectCell` renders `''` for both `undefined` (develop) and `null` (this branch), so this guard can't detect the change it's guarding. A `getSourceData()` assertion for the reverse shape in `raggedPaste.unit.ts` is red on develop and would cover it.

File: handsontable/src/plugins/copyPaste/__tests__/copyPaste.unit.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
+import Handsontable from '../../../index';
+import { registerCellType, TextCellType } from '../../../cellTypes';
+import { registerPlugin } from '../../registry';
+import { CopyPaste } from '../copyPaste';
+
+registerCellType(TextCellType);
+registerPlugin(CopyPaste);
+
+describe('CopyPaste ragged clipboard', () => {
```
Comment (Haxikowy) [inline handsontable/src/plugins/copyPaste/__tests__/copyPaste.unit.ts:9]: The filename matches no source module; the convention in this directory is module-named (`copyableRanges.unit.ts` for `copyableRanges.ts`). Case 1 also duplicates the plain-text E2E at roughly a second of jsdom. Cases 2 and 3 carry the value E2E can't reach.

**Review by Haxikowy — COMMENTED** [review]
#7361, cited in the description as possibly related, is not fixed and still reproduces here: `data: [['first'], ['alpha','beta','cookie','delta']]` with `minSpareCols: 1` loses `delta`, because the load path derives its column count from row 0 the same way. It is already closed as completed. Worth either dropping from the description or saying plainly that it isn't covered.

The docs checkbox is unchecked and #7389 carries the `Docs: Content` label. Nothing in `docs/content/guides/cell-features/clipboard/clipboard.md` states the old first-row-wins rule, so no text contradicts the code, but the new rule (widest row wins, short rows blank the cells they cover) is undocumented, as is `pasteMode: 'shift_down' | 'shift_right'` now shifting by the widest width.

File: handsontable/src/utils/parseTable.ts (hunk 1 above)
Comment (budnix) [reply handsontable/src/utils/parseTable.ts]: Nice catch — that footer case really did blow the width up to 20. Fixed in `84a613252b`, a row's last cell now takes just the one slot it lands in, with a test for it.

File: handsontable/src/utils/parseTable.ts (hunk 2 above)
Comment (budnix) [reply handsontable/src/utils/parseTable.ts]: Good spot — applied your `spannedEnd` clamp in `84a613252b`. Thanks for measuring it.

File: handsontable/src/utils/parseTable.ts (hunk 3 above)
Comment (budnix) [reply handsontable/src/utils/parseTable.ts]: You're right, matching the fill loop is the correct rule — switched to counting only `td` in `84a613252b`. It also cleared up the phantom column a `thead` colspan was adding.

File: handsontable/src/utils/parseTable.ts (hunk 4 above)
Comment (budnix) [reply handsontable/src/utils/parseTable.ts]: Fair point, but they lined up before only because the extra column was being dropped. Padding with `undefined` changes nothing since `getColHeader` falls back to the auto label anyway — and the `td`-only fix above removed the `thead`-wider case.

File: handsontable/src/plugins/copyPaste/copyPaste.ts (hunk 5 above)
Comment (budnix) [reply handsontable/src/plugins/copyPaste/copyPaste.ts:708]: Good point — the hooks describing a different paste than the grid was the real problem here. Padded before `beforePaste` in `84a613252b`; I kept the coercion as a guard, since a hook can still hand back a ragged array.

File: handsontable/src/plugins/copyPaste/copyPaste.ts (hunk 6 above)
Comment (budnix) [reply handsontable/src/plugins/copyPaste/copyPaste.ts:637]: Agreed, that was untested — added a repeat-fill case in `84a613252b` and called the change out in the changelog.

File: .changelogs/7389.json (hunk 7 above)
Comment (budnix) [reply .changelogs/7389.json:3]: Yep, my mistake — the title only covered half of it. Added a second `changed` entry in `84a613252b` for the blanking, the `null` fields and the repeat-fill width.

File: tests/e2e/clipboard.spec.ts (hunk 8 above)
Comment (budnix) [reply tests/e2e/clipboard.spec.ts]: Good catch, that assertion proved nothing either way. Replaced it in `84a613252b` with a `getSourceData()` case for the reverse shape.

File: handsontable/src/plugins/copyPaste/__tests__/copyPaste.unit.ts (hunk 9 above)
Comment (budnix) [reply handsontable/src/plugins/copyPaste/__tests__/copyPaste.unit.ts:9]: Agreed — renamed it to `copyPaste.unit.ts` in `84a613252b` and dropped the case the E2E already covers.

File: handsontable/src/utils/parseTable.ts (hunk 1 above)
Comment (Haxikowy) [reply handsontable/src/utils/parseTable.ts]: The footer case is fixed, but `cell === lastCell ? 1 : colSpan` under-reserves a cell that spans both ways. On `84a613252`, pasting this into a 4-column grid:

```html
<table><tbody><tr><td rowspan="2" colspan="2">A</td></tr><tr><td>B</td><td>C</td></tr></tbody></table>
```

gives `[["A",null,null,"d"],[null,null,"B","h"]]`. `C` is gone. `countCols` says 3, the fill loop needs 4, so `dataArr[row].findIndex(v => v === undefined)` returns -1 and the value goes to index -1. `dfc2d6a0c` pasted the same clipboard as 4 columns with both cells.

The rule also leaves the output ragged: `<tr><td>a</td></tr><tr><td>b</td><td colspan="3">wide</td></tr>` now returns row lengths `[2,4]`, with `mergeCells` colspan 3 pointing past `countCols`.

Counting the full `colSpan` and clamping the fill loop's merge write and the `mergeCells` colspan to `countCols` handles the footer row without the -1 write and keeps the rows rectangular. Either way, worth a case asserting nothing is dropped when the first row is a rowspan plus colspan cell.

File: handsontable/src/plugins/copyPaste/copyPaste.ts (hunk 5 above)
Comment (Haxikowy) [reply handsontable/src/plugins/copyPaste/copyPaste.ts:708]: Padding before the hook is the right place, but it squares off to the widest *physical* row, and the parser has already expanded the merged footer row to 20 entries. The grid writes 3 columns while the hook sees 20:

```
// 84a613252, paste of <tr><td>A1<td>B1<td>C1</tr><tr><td colspan="20">footer</tr>
parsed row lengths      [3, 20]
beforePaste row lengths [20, 20]
grid                    [["A1","B1","C1"], ["footer", null, null]]
```

So `afterPaste` reports 17 `null` columns the grid never wrote, which is the same mismatch with the sign flipped: an integrator syncing from the hook now writes phantom columns instead of missing real ones. Clamping the parser's rows to `countCols`, or padding once the paste width is settled, lines the two back up.

File: handsontable/src/utils/parseTable.ts (hunk 1 above)
Comment (budnix) [reply handsontable/src/utils/parseTable.ts]: You're right, the reservation had to stay the full span even where the row itself could spare it — fixed in `11d256e5ec`, with the rowspan-plus-colspan case tested. The ragged rows are gone too, by trimming the merge write rather than counting the full span, which would have brought the footer bug back.

File: handsontable/src/plugins/copyPaste/copyPaste.ts (hunk 5 above)
Comment (budnix) [reply handsontable/src/plugins/copyPaste/copyPaste.ts:708]: Good spot — the padding was squaring off rows the parser had already stretched. Trimming the merge write in `11d256e5ec` fixes it at the source, so the hook now reports 3 columns, same as the grid.

**Review by Haxikowy — APPROVED** [review]
